File: app/src/main/java/com/rokudo/xpense/data/repositories/BankApiRepo.java
package com.rokudo.xpense.data.repositories;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.rokudo.xpense.data.retrofit.GetDataService;
import com.rokudo.xpense.data.retrofit.RetrofitClientInstance;
import com.rokudo.xpense.data.retrofit.models.AccountDetails;
import com.rokudo.xpense.data.retrofit.models.Balances;
import com.rokudo.xpense.data.retrofit.models.DeleteResponse;
import com.rokudo.xpense.data.retrofit.models.EndUserAgreement;
import com.rokudo.xpense.data.retrofit.models.Institution;
import com.rokudo.xpense.data.retrofit.models.Requisition;
import com.rokudo.xpense.data.retrofit.models.RequisitionsResult;
import com.rokudo.xpense.data.retrofit.models.Token;
import com.rokudo.xpense.data.retrofit.models.TransactionsResponse;
import com.rokudo.xpense.utils.DatabaseUtils;
import com.rokudo.xpense.utils.NordigenUtils;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class BankApiRepo {
    private static final String TAG = "BankApiRepo";

    private static BankApiRepo instance;
    GetDataService service = RetrofitClientInstance.geInstance().create(GetDataService.class);
    private final MutableLiveData<Token> tokenMutableLiveData = new MutableLiveData<>();
    private String requisitionError;
    private boolean EUA_expired = false;
    //    private final MutableLiveData<List<Institution>> institutionLiveData;
//    private final MutableLiveData<AccountDetails> accountDetailsMutableLiveData;
    private final MutableLiveData<Balances> balancesMutableLiveData;
//    private final MutableLiveData<TransactionsResponse> accountTransactionsLiveData;

    public BankApiRepo() {
//        this.institutionLiveData = new MutableLiveData<>();
        this.balancesMutableLiveData = new MutableLiveData<>();
//        this.accountTransactionsLiveData = new MutableLiveData<>();
//        this.accountDetailsMutableLiveData = new MutableLiveData<>();
    }

    public String getRequisitionError() {
        return requisitionError;
    }

    public boolean isEUA_expired() {
        return EUA_expired;
    }

    public static BankApiRepo getInstance() {
        if (instance == null) {
            instance = new BankApiRepo();
        }
        return instance;
    }

    public MutableLiveData<Token> getToken() {
        service.getToken(NordigenUtils.NORDIGEN_SECRET_KEY_ID, NordigenUtils.NORDIGEN_SECRET_KEY)
                .enqueue(new Callback<Token>() {
                    @Override
                    public void onResponse(@NonNull Call<Token> call, @NonNull Response<Token> response) {
                        Log.d(TAG, "onResponse: ");
                        if (response.isSuccessful()) {
                            if (response.body() != null) {
                                NordigenUtils.TOKEN_VAL = response.body().getAccess();
                                tokenMutableLiveData.setValue(response.body());
                            }
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<Token> call, @NonNull Throwable t) {
                        Log.e(TAG, "onFailure: " + t.getMessage());
                    }
                });

        return tokenMutableLiveData;
    }

    public MutableLiveData<String> refreshToken(String token) {
        MutableLiveData<String> tokenLiveData = new MutableLiveData<>();

        service.refreshToken(token)
                .enqueue(new Callback<String>() {
                    @Override
                    public void onResponse(@NonNull Call<String> call, @NonNull Response<String> response) {
                        Log.d(TAG, "onResponse: ");
                        if (response.isSuccessful()) {
                            if (response.body() != null) {
                                tokenLiveData.setValue(response.body());
                            }
                        } else {
                            tokenLiveData.setValue(getErrorMessage(response.errorBody()));
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<String> call, @NonNull Throwable t) {
                        Log.e(TAG, "onFailure: " + t.getMessage());
                        tokenLiveData.setValue(null);
                    }
                });

        return tokenLiveData;
    }

    public MutableLiveData<List<Institution>> getInstitutionList() {
        MutableLiveData<List<Institution>> institutionLiveData = new MutableLiveData<>();

        service.getAllInstitutions()
                .enqueue(new Callback<List<Institution>>() {
                    @Override
                    public void onResponse(@NonNull Call<List<Institution>> call, @NonNull Response<List<Institution>> response) {
                        Log.d(TAG, "onResponse: ");
                        if (response.isSuccessful()) {
                            if (response.body() != null) {
                                institutionLiveData.setValue(response.body());
                            }
                        } else {
                            institutionLiveData.setValue(null);
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<List<Institution>> call, @NonNull Throwable t) {
                        Log.e(TAG, "onFailure: " + t.getMessage());
                        institutionLiveData.setValue(null);
                    }
                });

        return institutionLiveData;
    }

    public MutableLiveData<EndUserAgreement> createEUA(String institutionId) {
        MutableLiveData<EndUserAgreement> agreementMutableLiveData = new MutableLiveData<>();

        List<String> scopeList =
                new ArrayList<>(Arrays.asList("balances", "details", "transactions"));
        service.createEUA(institutionId, scopeList)
                .enqueue(new Callback<EndUserAgreement>() {
                    @Override
                    public void onResponse(@NonNull Call<EndUserAgreement> call, @NonNull Response<EndUserAgreement> response) {
                        Log.d(TAG, "onResponse: ");
                        if (response.isSuccessful()) {
                            Log.d(TAG, "onResponse: ");
                            agreementMutableLiveData.setValue(response.body());
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<EndUserAgreement> call, @NonNull Throwable t) {
                        Log.e(TAG, "onFailure: " + t.getMessage());
                        agreementMutableLiveData.setValue(null);
                    }
                });

        return agreementMutableLiveData;
    }

    public MutableLiveData<Requisition> createRequisition(String institutionID, String EUA_ID, Boolean account_selection) {
        MutableLiveData<Requisition> requisitionMutableLiveData = new MutableLiveData<>();

        service.createRequisition(institutionID,
                        "https://xpense/launch",
                        EUA_ID,
                        "EN",
                        DatabaseUtils.getCurrentUser().getPhoneNumber() + "_" + institutionID,
                        account_selection,
                        true)
                .enqueue(new Callback<Requisition>() {
                    @Override
                    public void onResponse(@NonNull Call<Requisition> call, @NonNull Response<Requisition> response) {
                        if (response.isSuccessful()) {
                            Log.d(TAG, "onResponse: success ");
                            requisitionError = null;
                            requisitionMutableLiveData.setValue(response.body());
                        } else {
                            Log.e(TAG, "onResponse: failed " + response.message());
                            requisitionError = getErrorMessage(response.errorBody());
                            requisitionMutableLiveData.setValue(null);
                        }

                    }

                    @Override
                    public void onFailure(@NonNull Call<Requisition> call, @NonNull Throwable t) {
                        Log.e(TAG, "onFailure: " + t.getMessage());
                        requisitionError = t.getMessage();
                        requisitionMutableLiveData.setValue(null);
                    }
                });

        return requisitionMutableLiveData;
    }

    public MutableLiveData<Requisition> getRequisitionDetails(String requisitionID) {
        MutableLiveData<Requisition> requisitionMutableLiveData = new MutableLiveData<>();

        service.getRequisitionById(requisitionID).enqueue(new Callback<Requisition>() {
            @Override
            public void onResponse(@NonNull Call<Requisition> call, @NonNull Response<Requisition> response) {
                if (response.isSuccessful()) {
                    Log.d(TAG, "onResponse: success ");
                    if (response.body() != null) {
                        requisitionMutableLiveData.setValue(response.body());
                    }

                } else {
                    Log.e(TAG, "onResponse: failed " + response.message());
                    requisitionMutableLiveData.setValue(null);
                }
            }

            @Override
            public void onFailure(@NonNull Call<Requisition> call, @NonNull Throwable t) {
                Log.e(TAG, "onFailure: " + t.getMessage());
                requisitionMutableLiveData.setValue(null);
            }
        });

        return requisitionMutableLiveData;
    }

    public MutableLiveData<AccountDetails> getAccountDetails(String account_id) {
        MutableLiveData<AccountDetails> accountDetailsMutableLiveData = new MutableLiveData<>();

        service.getAccountDetails(account_id).enqueue(new Callback<AccountDetails>() {
            @Override
            public void onResponse(@NonNull Call<AccountDetails> call, @NonNull Response<AccountDetails> response) {
                Log.d(TAG, "onResponse: ");
                if (response.isSuccessful()) {
                    AccountDetails accountDetails = response.body();
                    if (accountDetails != null) {
                        accountDetails.setAccount_id(account_id);
                    }
                    accountDetailsMutableLiveData.setValue(accountDetails);
                } else {
                    Log.e(TAG, "onResponse: " + getErrorMessage(response.errorBody()));
                    accountDetailsMutableLiveData.setValue(null);
                }
            }

            @Override
            public void onFailure(@NonNull Call<AccountDetails> call, @NonNull Throwable t) {
                Log.e(TAG, "onResponse: " + t.getMessage());
                accountDetailsMutableLiveData.setValue(null);
            }
        });

        return accountDetailsMutableLiveData;
    }

    public MutableLiveData<Balances> getAccountBalances(String account_id) {
//        MutableLiveData<Balances> balancesMutableLiveData = new MutableLiveData<>();

        service.getAccountBalances(account_id).enqueue(new Callback<Balances>() {
            @Override
            public void onResponse(@NonNull Call<Balances> call, @NonNull Response<Balances> response) {
                if (response.isSuccessful()) {
                    balancesMutableLiveData.setValue(response.body());
                } else {
                    balancesMutableLiveData.setValue(null);
                    Log.e(TAG, "onResponse: " + getErrorMessage(response.errorBody()));
                    if (response.code() == 409) {
                        EUA_expired = true;
                    }
                }
            }

            @Override
            public void onFailure(@NonNull Call<Balances> call, @NonNull Throwable t) {
                balancesMutableLiveData.setValue(null);
                Log.e(TAG, "onFailure: ", t);
            }
        });

        return balancesMutableLiveData;
    }

    public void setBalances(Balances balances) {
        balancesMutableLiveData.setValue(balances);
    }

    public MutableLiveData<TransactionsResponse> getAccountTransactions(String account_id, String date_from) {
        MutableLiveData<TransactionsResponse> accountTransactionsLiveData = new MutableLiveData<>();
        service.getAccountTransactions(account_id, date_from)
                .enqueue(new Callback<TransactionsResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<TransactionsResponse> call, @NonNull Response<TransactionsResponse> response) {
                        if (response.isSuccessful()) {
                            Log.d(TAG, "onResponse: " + response.body());
                            requisitionError = null;
                            accountTransactionsLiveData.setValue(response.body());
                        } else {
                            Log.e(TAG, "onResponse: " + response.message());
                            requisitionError = getErrorMessage(response.errorBody());
                            accountTransactionsLiveData.setValue(null);
                            if (response.code() == 409) {
                                EUA_expired = true;
                            }
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<TransactionsResponse> call, @NonNull Throwable t) {
                        Log.e(TAG, "onFailure: ", t);
                        requisitionError = t.getMessage();
                        accountTransactionsLiveData.setValue(null);
                    }
                });

        return accountTransactionsLiveData;
    }

    private String getErrorMessage(ResponseBody response) {
        if (response == null)
            return null;

        String errorMessage = null;
        try {
            JSONObject jsonObject = new JSONObject(response.string());
            errorMessage = jsonObject.toString();
            Log.e(TAG, "onResponse: " + jsonObject);
        } catch (IOException | JSONException e) {
            e.printStackTrace();
        }
        return errorMessage;
    }

    public MutableLiveData<RequisitionsResult> getAllRequisitions() {
        MutableLiveData<RequisitionsResult> requisitionsResultMutableLiveData = new MutableLiveData<>();

        service.getRequisitions().enqueue(new Callback<RequisitionsResult>() {
            @Override
            public void onResponse(@NonNull Call<RequisitionsResult> call, @NonNull Response<RequisitionsResult> response) {
                if (response.isSuccessful()) {
                    requisitionsResultMutableLiveData.setValue(response.body());
                } else {
                    requisitionsResultMutableLiveData.setValue(null);
                }
            }

            @Override
            public void onFailure(@NonNull Call<RequisitionsResult> call, @NonNull Throwable t) {
                Log.e(TAG, "onFailure: ", t);
            }
        });

        return requisitionsResultMutableLiveData;
    }

    public void deleteRequisition(String requisition_id) {
        service.deleteRequisition(requisition_id).enqueue(new Callback<DeleteResponse>() {
            @Override
            public void onResponse(@NonNull Call<DeleteResponse> call, @NonNull Response<DeleteResponse> response) {
                Log.d(TAG, "onResponse: " + response.message());
            }

            @Override
            public void onFailure(@NonNull Call<DeleteResponse> call, @NonNull Throwable t) {
                Log.e(TAG, "onFailure: ", t);
            }
        });
    }

    public void deleteEua(String eua_id) {
        service.deleteEua(eua_id).enqueue(new Callback<DeleteResponse>() {
            @Override
            public void onResponse(@NonNull Call<DeleteResponse> call, @NonNull Response<DeleteResponse> response) {
                Log.d(TAG, "onResponse: " + response.message());
            }

            @Override
            public void onFailure(@NonNull Call<DeleteResponse> call, @NonNull Throwable t) {
                Log.e(TAG, "onFailure: ", t);
            }
        });
    }
}

File: app/src/main/java/com/rokudo/xpense/data/repositories/BankApiRepo.kt
package com.rokudo.xpense.data.repositories

import android.util.Log
import androidx.lifecycle.MutableLiveData
import com.rokudo.xpense.data.retrofit.GetDataService
import com.rokudo.xpense.data.retrofit.RetrofitClientInstance
import com.rokudo.xpense.data.retrofit.models.*
import com.rokudo.xpense.utils.DatabaseUtils
import com.rokudo.xpense.utils.GoCardlessUtils
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class BankApiRepo {
    companion object {
        private const val TAG = "BankApiRepo"
        val instance by lazy { BankApiRepo() }
    }

    private val service: GetDataService = RetrofitClientInstance.getInstance().create(GetDataService::class.java)
    private val tokenMutableLiveData = MutableLiveData<Token>()
    private var requisitionError: String? = null
    private var euaExpired = false
    private val balancesMutableLiveData = MutableLiveData<Balances>()
    private val accountTransactionsLiveData = MutableLiveData<TransactionsResponse>()

    fun getToken(): MutableLiveData<Token> {
        service.getToken(GoCardlessUtils.GOCARDLESS_SECRET_KEY_ID, GoCardlessUtils.GOCARDLESS_SECRET_KEY)
            .enqueue(object : Callback<Token> {
                override fun onResponse(call: Call<Token>, response: Response<Token>) {
                    if (response.isSuccessful && response.body() != null) {
                        GoCardlessUtils.TOKEN_VAL = response.body()!!.access ?: ""
                        tokenMutableLiveData.value = response.body()
                    }
                }
                override fun onFailure(call: Call<Token>, t: Throwable) {
                    Log.e(TAG, "onFailure: ${t.message}")
                }
            })
        return tokenMutableLiveData
    }

    fun refreshToken(token: String): MutableLiveData<String> {
        val tokenLiveData = MutableLiveData<String>()
        service.refreshToken(token).enqueue(object : Callback<String> {
            override fun onResponse(call: Call<String>, response: Response<String>) {
                if (response.isSuccessful) {
                    tokenLiveData.value = response.body()
                } else {
                    tokenLiveData.value = getErrorMessage(response.errorBody())
                }
            }
            override fun onFailure(call: Call<String>, t: Throwable) {
                tokenLiveData.value = null
            }
        })
        return tokenLiveData
    }

    fun getInstitutionList(): MutableLiveData<List<Institution>> {
        val data = MutableLiveData<List<Institution>>()
        service.getAllInstitutions().enqueue(object : Callback<List<Institution>> {
            override fun onResponse(call: Call<List<Institution>>, response: Response<List<Institution>>) {
                data.value = if (response.isSuccessful) response.body() else null
            }
            override fun onFailure(call: Call<List<Institution>>, t: Throwable) {
                data.value = null
            }
        })
        return data
    }

    fun createEUA(institutionId: String): MutableLiveData<EndUserAgreement> {
        val data = MutableLiveData<EndUserAgreement>()
        service.createEUA(institutionId, listOf("balances", "details", "transactions"))
            .enqueue(object : Callback<EndUserAgreement> {
                override fun onResponse(call: Call<EndUserAgreement>, response: Response<EndUserAgreement>) {
                    data.value = if (response.isSuccessful) response.body() else null
                }
                override fun onFailure(call: Call<EndUserAgreement>, t: Throwable) {
                    data.value = null
                }
            })
        return data
    }

    fun createRequisition(institutionID: String, euaId: String, accountSelection: Boolean?): MutableLiveData<Requisition> {
        val data = MutableLiveData<Requisition>()
        service.createRequisition(
            institutionID, "https://xpense/launch", euaId, "EN",
            "${DatabaseUtils.currentUser?.phoneNumber}_$institutionID",
            accountSelection, true
        ).enqueue(object : Callback<Requisition> {
            override fun onResponse(call: Call<Requisition>, response: Response<Requisition>) {
                if (response.isSuccessful) {
                    requisitionError = null
                    data.value = response.body()
                } else {
                    requisitionError = getErrorMessage(response.errorBody())
                    data.value = null
                }
            }
            override fun onFailure(call: Call<Requisition>, t: Throwable) {
                requisitionError = t.message
                data.value = null
            }
        })
        return data
    }

    fun getRequisitionDetails(requisitionID: String): MutableLiveData<Requisition> {
        val data = MutableLiveData<Requisition>()
        service.getRequisitionById(requisitionID).enqueue(object : Callback<Requisition> {
            override fun onResponse(call: Call<Requisition>, response: Response<Requisition>) {
                data.value = if (response.isSuccessful) response.body() else null
            }
            override fun onFailure(call: Call<Requisition>, t: Throwable) {
                data.value = null
            }
        })
        return data
    }

    fun getAccountDetails(accountId: String): MutableLiveData<AccountDetails> {
        val data = MutableLiveData<AccountDetails>()
        service.getAccountDetails(accountId).enqueue(object : Callback<AccountDetails> {
            override fun onResponse(call: Call<AccountDetails>, response: Response<AccountDetails>) {
                if (response.isSuccessful) {
                    val details = response.body()
                    details?.account_id = accountId
                    data.value = details
                } else {
                    data.value = null
                }
            }
            override fun onFailure(call: Call<AccountDetails>, t: Throwable) {
                data.value = null
            }
        })
        return data
    }

    fun getAccountBalances(accountId: String): MutableLiveData<Balances> {
        if (GoCardlessUtils.TOKEN_VAL.isEmpty()) {
            service.getToken(GoCardlessUtils.GOCARDLESS_SECRET_KEY_ID, GoCardlessUtils.GOCARDLESS_SECRET_KEY)
                .enqueue(object : Callback<Token> {
                    override fun onResponse(call: Call<Token>, response: Response<Token>) {
                        if (response.isSuccessful && response.body() != null) {
                            GoCardlessUtils.TOKEN_VAL = response.body()!!.access ?: ""
                            tokenMutableLiveData.value = response.body()
                            getAccountBalances(accountId)
                        }
                    }
                    override fun onFailure(call: Call<Token>, t: Throwable) {
                        Log.e(TAG, "onFailure: ${t.message}")
                    }
                })
        } else {
            service.getAccountBalances(accountId).enqueue(object : Callback<Balances> {
                override fun onResponse(call: Call<Balances>, response: Response<Balances>) {
                    if (response.isSuccessful) {
                        val currentVal = balancesMutableLiveData.value
                        val newBody = response.body()
                        if (currentVal == null || newBody?.balances == null ||
                            currentVal.balances?.size != newBody.balances?.size
                        ) {
                            balancesMutableLiveData.value = newBody
                        } else {
                            var updateNeeded = false
                            newBody.balances?.forEachIndexed { i, balance ->
                                try {
                                    val old = currentVal.balances?.get(i)?.balanceAmount?.get("amount")
                                    val new_ = balance.balanceAmount?.get("amount")
                                    if (old != null && old != new_) updateNeeded = true
                                } catch (_: Exception) {}
                            }
                            if (updateNeeded) balancesMutableLiveData.value = newBody
                        }
                    } else {
                        balancesMutableLiveData.value = null
                        if (response.code() == 409) euaExpired = true
                    }
                }
                override fun onFailure(call: Call<Balances>, t: Throwable) {
                    balancesMutableLiveData.value = null
                }
            })
        }
        return balancesMutableLiveData
    }

    fun setBalances(balances: Balances?) {
        balancesMutableLiveData.value = balances
    }

    fun getAccountTransactions(accountId: String, dateFrom: String): MutableLiveData<TransactionsResponse> {
        service.getAccountTransactions(accountId, dateFrom).enqueue(object : Callback<TransactionsResponse> {
            override fun onResponse(call: Call<TransactionsResponse>, response: Response<TransactionsResponse>) {
                if (response.isSuccessful) {
                    requisitionError = null
                    accountTransactionsLiveData.value = response.body()
                } else {
                    requisitionError = getErrorMessage(response.errorBody())
                    accountTransactionsLiveData.value = null
                    if (response.code() == 409) euaExpired = true
                }
            }
            override fun onFailure(call: Call<TransactionsResponse>, t: Throwable) {
                requisitionError = t.message
                accountTransactionsLiveData.value = null
            }
        })
        return accountTransactionsLiveData
    }

    private fun getErrorMessage(response: ResponseBody?): String? {
        if (response == null) return null
        return try {
            val json = JSONObject(response.string())
            json.toString()
        } catch (e: Exception) {
            null
        }
    }

    fun getAllRequisitions(): MutableLiveData<RequisitionsResult> {
        val data = MutableLiveData<RequisitionsResult>()
        service.getRequisitions().enqueue(object : Callback<RequisitionsResult> {
            override fun onResponse(call: Call<RequisitionsResult>, response: Response<RequisitionsResult>) {
                data.value = if (response.isSuccessful) response.body() else null
            }
            override fun onFailure(call: Call<RequisitionsResult>, t: Throwable) {
                Log.e(TAG, "onFailure: ", t)
            }
        })
        return data
    }

    fun deleteRequisition(requisitionId: String) {
        service.deleteRequisition(requisitionId).enqueue(object : Callback<DeleteResponse> {
            override fun onResponse(call: Call<DeleteResponse>, response: Response<DeleteResponse>) {}
            override fun onFailure(call: Call<DeleteResponse>, t: Throwable) {}
        })
    }

    fun deleteEua(euaId: String) {
        service.deleteEua(euaId).enqueue(object : Callback<DeleteResponse> {
            override fun onResponse(call: Call<DeleteResponse>, response: Response<DeleteResponse>) {}
            override fun onFailure(call: Call<DeleteResponse>, t: Throwable) {}
        })
    }

    fun getRequisitionError(): String? = requisitionError
    fun isEUA_expired(): Boolean = euaExpired
}

